torch/fx/experimental/Partitioner.py

```diff
@@ -126,22 +174,46 @@ def __init__(self) -> None:
         self.partitions: List[Partition] = []
         self.devices: List[Dict[str, Union[str, float]]] = []
         self.node_to_partitions: Dict[Node, List[int]] = {}
+        self.partition_to_used_mem: Dict[Partition, int] = {}
+        self.orphans: List[Node] = []
+        self.args_mapping: Dict[Node, List[Node]] = []
+        self.users_mapping: Dict[Node, List[Node]] = []
 
-    def partition_graph(self, fx_module: GraphModule, devices: List[Dict[str, Union[str, float]]]) -> DAG:
+    def partition_graph(self, fx_module: GraphModule, torch_module: torch.nn.Module, devices: List[Tuple[str, float, int]]) -> Tuple[DAG, GraphModule]:
         """
-        Given the fx_module and devices, find the partitions, do the partitions,
+        Given the fx_module and devices (a list of namedtuple ('device', ['name', 'available_mem_bytes'])),
+        find the partitions, and then do the partitions,
         and then return a dictionary representing the DAG of partitions
         """
         self.graph_module = fx_module
         self.devices = devices
+        self.torch_module = torch_module
         # Create a dummy partition as root.
         self.root_partition = self.create_partition()
-        # So far, the whole could fit into one device since we assume the memory is unlimited.
-        # TODO: Check the available memory size for each device and see if one device is enough or multiple partitions are needed.
-        self.find_single_partition()
-        self.do_partition()
+        if len(self.devices) == 0:
+            raise RuntimeError('No devices')
+        available_mem = self.devices[0].available_mem_bytes
+        # Calculate total size of the graph
+        total_size_of_graph = 0
+        for node in self.graph_module.graph.nodes:
+            if node.op == 'output':
+                break
+            total_size_of_graph += node.size.total_size
+        if total_size_of_graph <= available_mem:
+            self.find_single_partition()
+        elif total_size_of_graph > len(self.devices) * available_mem:
+            raise RuntimeError('Devices have no enough memory for the module')
+        else:
+            self.size_based_partition(available_mem)
+        # Check if enought devices are provided for all partitions
+        if len(self.partitions) - 1 > len(self.devices):
+           raise RuntimeError('Lack of Devices')
+        module_with_submodules = self.do_partition()
+        # The DAG contains DAGNodes with info of each partition's input nodes, output nodes
+        # and how partitions are connected.
         dag = self.dump_partition_DAG()
-        return dag
+        ret = namedtuple('ret', ['dag', 'module_with_submodules'])
```

Comment:
The dag has all info on the nodes of base graph module. The module_with_submodules is the new graph module that is generated by split_module